erts/emulator/beam/utils.c

```diff
@@ -5206,48 +5198,33 @@ static void *erts_qsort_get_item(void *base, size_t item_size, size_t index)
  * means that the code has to follow some restrictions. See note about
  * YCF near the top of the file for more information.
  */
-static erts_qsort_partion_array_result
-erts_qsort_partion_array(void *base,
+static byte*
+erts_qsort_partion_array(byte *base,
                          size_t nr_of_items,
                          size_t item_size,
                          erts_void_ptr_cmp_t compare,
                          int extra_seed)
 {
-    size_t last_part_start = nr_of_items;
-    size_t pivot_part_start = 0;
-    size_t pivot_part_end = 1;
+    byte* second_part_start = base + (nr_of_items * item_size);
+    byte* curr = base + item_size;
     size_t pivot_index =
         (size_t)erts_sched_local_random_hash_64_to_32_shift((Uint64)extra_seed +
                                                             (Uint64)((UWord)compare) +
                                                             (Uint64)((UWord)base) -
                                                             (Uint64)nr_of_items) % nr_of_items;
     /* Move pivot first */
-    erts_qsort_swap(base, item_size, 0, pivot_index);
-    while (pivot_part_end != last_part_start) {
-        const void *item =
-            erts_qsort_get_item(base, item_size, pivot_part_end);
-        int compare_res =
-            compare(item, erts_qsort_get_item(base, item_size, pivot_part_end - 1));
-        if (compare_res == 0) {
-            /* Include in pivot part */
-            pivot_part_end++;
-        } else if (compare_res < 0) {
-            /* Move pivot part one step to the right */
-            erts_qsort_swap(base, item_size, pivot_part_start, pivot_part_end);
-            pivot_part_start++;
-            pivot_part_end++;
+    erts_qsort_swap(item_size, base, base + pivot_index*item_size);
+    while (curr != second_part_start) {
```

Comment:
Perhaps there is some better way to optimize the partition function so we can keep the pivot partition and still reduce the number of swaps?

If we skip the pivot partition, we should name the erts_qsort function something else and place it in `erl_map.c` because no pivot partition can lead to O(N^2) run time when all sorted items are equal.

Reply:
Fixed